File: core/src/com/alexpi/awesometanks/screens/upgrades/UpgradeTable.java
package com.alexpi.awesometanks.screens.upgrades;

import com.alexpi.awesometanks.screens.widget.GameProgressBar;
import com.alexpi.awesometanks.screens.widget.Styles;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.scenes.scene2d.ui.Cell;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.utils.Align;

public class UpgradeTable extends Table {

    private GameProgressBar bar;
    private Label nameLabel;
    private TextButton buy;
    private int price;
    public UpgradeTable(AssetManager assetManager, String name, float value, float max, int price, boolean vertical){
        this.price = price;
        nameLabel = new Label(name, Styles.getLabelStyleSmall(assetManager));
        nameLabel.setAlignment(Align.center);
        buy = new TextButton("Buy "+price+" $", Styles.getTextButtonStyleSmall(assetManager));
        bar = new GameProgressBar(assetManager, max,0f, vertical);
        bar.setValue(value);
        Cell<GameProgressBar> barCell = add(bar);
        if(vertical) barCell.expandY().fillY();
        else barCell.fillX();
        barCell.row();
        add(nameLabel).padTop(8).padBottom(8).row();
        add(buy).row();

    }
    public UpgradeTable(AssetManager assetManager, String name, float value, float max, int price){
        this(assetManager, name, value, max, price, false);
    }

    public void changePrice(int price){this.price=price;buy.setText("Buy " + price + " $");}

    public boolean canBuy(int money){
        return (money >=price && bar.getValue() < bar.getMaxValue());
    }

    public int getPrice(){return price;}
    public TextButton getBuyButton() {return buy;}
    public String getName(){return nameLabel.getText().toString();}
    public void increaseValue(int i) {
        if(bar.getValue() + i > bar.getMaxValue()){
            bar.setValue(bar.getMaxValue());
        } else bar.setValue(bar.getValue()+i);
    }
    public void setValue(float value){bar.setValue(value);}
    public int getValue(){return (int) bar.getValue();}

    public boolean isMaxValue(){return bar.getValue() == bar.getMaxValue();}

    public float getMaxValue() {return bar.getMaxValue();}
}

File: core/src/com/alexpi/awesometanks/screens/upgrades/UpgradeTable.kt
package com.alexpi.awesometanks.screens.upgrades

import com.alexpi.awesometanks.screens.widget.FloatProgressBar
import com.alexpi.awesometanks.screens.widget.Styles
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.utils.Align
import ktx.actors.onClick

class UpgradeTable(
    assetManager: AssetManager,
    upgradeName: String,
    currentLevel: Float,
    maxLevel: Float,
    upgradePrice: Int,
    vertical: Boolean = false
) : Table() {
    private val bar: FloatProgressBar
    private val nameLabel = Label(upgradeName, Styles.getLabelStyleSmall(assetManager))
    private val buyButton: TextButton

    var upgradePrice: Int = upgradePrice
        set(value) {
            field = value
            buyButton.setText("Buy $value $")
        }

    var onBuyClick: (() -> Unit)? = null

    init {
        nameLabel.setAlignment(Align.center)
        buyButton = TextButton("Buy $upgradePrice $", Styles.getTextButtonStyleSmall(assetManager))
        bar = FloatProgressBar(
            assetManager,
            maxLevel,
            0f,
            vertical
        )
        bar.currentValue = currentLevel
        if(bar.currentValue == maxLevel) buyButton.isVisible = false
        val barCell = add(bar)
        if (vertical) barCell.expandY().fillY()
        else barCell.fillX()
        barCell.row()
        add(nameLabel).padTop(8f).padBottom(8f).row()
        add(buyButton).row()

        buyButton.onClick {
            onBuyClick?.invoke()
        }
    }

    fun canBuy(money: Int): Boolean {
        return (money >= upgradePrice && bar.currentValue < bar.maxValue)
    }

    override fun getName(): String {
        return nameLabel.text.toString()
    }

    var upgradeLevel: Float
        get() = bar.currentValue
        set(value) {
            if(value > bar.maxValue) {
                bar.currentValue = bar.maxValue
                buyButton.isVisible = false
            } else {
                bar.currentValue = value
                buyButton.isVisible = true
            }
        }

    val isMaxLevel: Boolean
        get() = bar.currentValue == bar.maxValue

}
